Start sympy.codegen module with ast.py from symcc

asmeurer opened the pull request:
Symcc is from https://github.com/jcrist/symcc.

CC @jcrist 

TODO:
- [x] Merge stuff from codeprinters (Assignment)
- [ ] Port ccode and fcode printers from symcc
- [ ] Routine stuff from symcc?
- [ ] Fix test_args

asmeurer committed:
Add (experimental) CodeBlock object

asmeurer committed:
Add a basic __new__ with some type checking to CodeBlock

asmeurer committed:
Add a basic cse() method to CodeBlock()

As it currently stands, it is incorrect, because the assignments from the cse
are placed first but they might reference variables that aren't defined yet.

asmeurer committed:
Merge branch 'master' into codeblock

asmeurer committed:
Add topological_sort constructor to CodeBlock

This lets you construct CodeBlock instances with unsorted assignments
(variables used before they are defined). This will be needed for cse().

This was added as a class method constructor because I plan to add a check to
the normal constructor for undefined variables.

asmeurer committed:
Add left_hand_sides and right_hand_sides properties to CodeBlock

asmeurer committed:
CodeBlock.cse only takes the cse of the right-hand sides

We don't want to replace a duplicate Assignment with a variable (even though
they aren't allowed yet).

asmeurer committed:
CodeBlock.cse now correctly returns a topologically sorted block

In other words, variables are not used before they are defined.

jcrist committed:
Start sympy.codegen module with ast.py from symcc

Symcc is from https://github.com/jcrist/symcc.

asmeurer commented:
I plan to do some cleanups and improvements here. Don't merge until ready (but feel free to review at any time).

asmeurer committed:
Add sympy/codegen/__init__.py (empty for now)

asmeurer committed:
Add codegen module to setup.py

asmeurer committed:
Fix codegen doctests

asmeurer committed:
Use string_types instead of str in sympy.codegen.ast

asmeurer committed:
For AugAssign it will be better to use the string form of the operator

That way you do not accidentally write AugAssign(x, AddOp, y) instead of
AugAssign(x, AddOp(), y) (which is what the doctest originally did).

It might also be a good idea to only expose the singleton objects, or give
they a different name from their classes, as is done with singletons in the
core.

asmeurer committed:
Move Assignment from sympy.printing.codeprinter to sympy.codegen.ast

The old import is kept for backwards compatibility but all imports in SymPy
have been changed.

This replaces the identical Assign object that was there from symcc.

This required some moving of imports inside functions to avoid circular import
issues.

asmeurer commented on sympy/codegen/ast.py:
```diff
@@ -0,0 +1,627 @@
+# tokens.
+
+
+class NativeOp(with_metaclass(Singleton, Basic)):
```
Is Singleton really needed here? Just because an object is atomic (no args) doesn't mean it needs to be singletonized. `op == AddOp()` or `isinstance(op, AddOp)` should work regardless, because of the way equality is defined on Basic. The only thing Singleton enables is `is` comparison, which is a bit of an antipattern unless it's needed for performance or memory purposes (which it really isn't here).

asmeurer committed:
De-singletonize classes in sympy.codegen.ast

Just because a class is atomic (no args) doesn't mean it needs to be a
singleton. == from Basic will work just fine, and so will isinstance checks. S
shouldn't be cluttered up with relatively little used classes that have little
benefit from the performance or memory gains of being singletonized.

asmeurer commented on sympy/codegen/ast.py:
```diff
@@ -0,0 +1,627 @@
+
+
+class NativeOp(with_metaclass(Singleton, Basic)):
+    """Base type for native operands."""
```
What does "native" mean in this context?

jcrist replied:
I think I just added that to distinguish it from sympy's operators. They were only used to avoid having strings as args in the augmented assignment classes (looks like you already figured that out though :)). I'm not sure if the augmented assignment is even needed though, `a += b` and `a = a + b` should result in the same compiled code.

In general, I haven't thought about this in > 1 year, so you probably have better ideas on how to go about doing this in sympy than I do.

asmeurer replied:
> I'm not sure if the augmented assignment is even needed though, a += b and a = a + b should result in the same compiled code.

That's a good point. It might make for a good abstraction (like when writing a sum as an accumulator). I'll leave them in here for now.

asmeurer committed:
sympy.codegen.ast: Make separate Assignment subclasses for each augmented assignment

This is a much cleaner design than before. We can now re-use the Assignment
constructor and its type checking (which was previously duplicated in
AugAssign). Also being subclasses of Relational will make them work better
with other things.

There are no longer the confusing operator nodes (operators
in SymPy expression trees are generally the functions, not the nodes
themselves).

Also use longer names (AugmentedAssignment), matching the existing
Assignment.

AugAssign has been renamed as aug_assign, a helper function to create
AddAugmentedAssignment, SubAugmentedAssignment, etc.

asmeurer committed:
Remove _sympystr from AugmentedAssignment

The default printer on Relational already does the right thing.

asmeurer committed:
Fix string and pretty printing of AugmentedAssignment objects

str printing should always be valid Python (+= not only isn't valid in
expressions, which AddAugmentedAssignment could go in for languages like C
where this works, but it doesn't recreate the object being printed).  The
pretty printers print the assignments using += and so on.

asmeurer committed:
Update class chart at the top of sympy/codegen/ast.py

bjodah commented:
I think it looks good overall, but it is hard to know if the design has the needed capabilities before one has tried it for something non-trivial.

The only thing that stood out to me was the `For` class, it is very "pythonic", but perhaps it's not meant to map directly to a C-style for loop (or even a Fortran for loop)? I guess a `While` class could be useful as well? (e.g. describing fixed-point iteration with a convergence criterion).

jcrist commented:
> The only thing that stood out to me was the For class, it is very "pythonic", but perhaps it's not meant to map directly to a C-style for loop (or even a Fortran for loop)?

SymPy has `Range` objects, which would map exactly to fortran/c style loops. Check out the tests here: https://github.com/jcrist/symcc/blob/master/symcc/printers/tests/test_ccode.py#L203.

I do agree that it's hard to know if a design is good until it's been used for something non-trivial though.

asmeurer commented:
Well my short-term goal is to be able to generate code for Sum (specifically dot products on variable length vectors) https://github.com/sympy/sympy/issues/10194. For that I think `For` and `AddAgumentedAssignment` will come in handy. In that case, the user won't actually use `For` directly. The code printer for Sum will map to it. This at least saves us from rewriting for loops in several code printers, which is already a gain. `While` would probably be useful, although I don't have any specific use-cases for it yet. 

Another thing that I see now that's missing is a higher level construct for blocks of statements. `For` from symcc just stores the body in a `Tuple`. I started something [earlier on this](https://github.com/sympy/sympy/compare/master...asmeurer:codeblock). I spent a bit of time working out cse, which is a neat trick (it lets you perform cse on a code AST expression and get a code AST expression back, which is actually nontrivial to do right if you care about variables being defined before they are used), but perhaps a bit of a sidetracked goal. I'll probably port my codeblock stuff to this branch. 

I definitely am agreeing when it comes to the other things, especially the abstractions surrounding types. I have no idea what the "right" way to abstract types should be, and I'm a little hesitant to guess without a good real-world example to guide me.

asmeurer committed:
Merge branch 'codeblock' into symcc

asmeurer committed:
Move CodeBlock from sympy.printing.codeprinter to sympy.codegen.ast

asmeurer committed:
Add some tests for CodeBlock

asmeurer committed:
Add some basic docstrings/doctests for CodeBlock and its methods

asmeurer commented on sympy/codegen/ast.py:
```diff
@@ -0,0 +1,732 @@
+        # De-enumerate the result
+        return cls(*list(zip(*ordered_assignments))[1])
+
+    def cse(self, symbols=None, optimizations=None, postprocess=None,
```
For this to be useful in a C setting it needs to somehow return the new variables to the user so that they can be initialized properly.

asmeurer commented on sympy/codegen/ast.py:
```diff
@@ -0,0 +1,732 @@
+    Useful methods on this object are
+
+    ``left_hand_sides``: Tuple of left-hand sides, in order.
+    ``left_hand_sides``: Tuple of right-hand sides, in order.
```
Will these make sense when the code block can hold more than just assignments?

asmeurer commented on sympy/codegen/ast.py:
```diff
@@ -0,0 +1,732 @@
+        return obj
+
+    @classmethod
+    def topological_sort(cls, assignments):
```
This probably does the wrong things for augmented assignments.

bjodah replied:
If you expect this not to work with augmented assignments perhaps we should raise a `NotImplementedError` or at least mention in the docstring that the behaviour is undefined?

asmeurer replied:
Ah, I forgot about this TODO. I should definitely fix this.

asmeurer committed:
Move restrictions in CodeBlock regarding Assignments to CodeBlock.cse

The restrictions that every argument be an assignment and that duplicate
assignments are disallowed are only needed because of cse. We want CodeBlock
to be useable in a more general setting, though.

Also add an error for AugmentedAssignments, since they don't work the same way
as Assignments (the variable must already be assigned, meaning duplicate
assignments must be allowed).

These are all NotImplementedErrors. They can work, but some additional things
need to be implemented.

asmeurer committed:
Merge branch 'master' into symcc

asmeurer committed:
Small changes to sympy.codegen.ast.For

- Add doctest.
- Handle iter being a list.
- Store the body as a CodeBlock instead of a Tuple.

asmeurer committed:
Fix For doctest

asmeurer committed:
Scrap the rest of sympy.codegen.ast for now

I don't need any of this stuff just yet, and I'm unsure if the way that it was
done is the best way to do it, and am wary of guessing without a good
use-case.

Stuff removed:

Datatype stuff: I'm particularly worried about this, and I am trying to avoid
doing type stuff
for now (side note: it seems to me rather that the datatypes should be
classes, like Double(x)).

Variable and its subclasses: the whole point of that was to attach datatypes
to variables (and I also don't get the InOutAssignment stuff; that seems too
Fortran specific and that it ought to be abstracted in a different way or even
done automatically).

FunctionDef: functions seem useful, but I also didn't port the routine stuff
from symcc either. I don't know what the most useful set of language agnostic
abstractions for these things are. We'll probably need something eventually,
but for now CodeBlock is a decent way to combine code expressions.

Import: This only makes sense if we can codegen an entire file, which, without
the other stuff, isn't going to happen any time soon.

Declare: Only useful if we deal with datatypes.

Return: Not useful without FunctionDef. I'm also not sure if this is the right
abstraction.

asmeurer commented:
I just dropped a bunch of stuff I'm unsure about. It's still there in symcc if we decide we want it later, but for now, what remains is enough for what I want to do.

asmeurer commented:
(see the commit message(s) for more information)

asmeurer committed:
Remove unused import

asmeurer committed:
Add sympy.codegen classes to test_args

asmeurer committed:
Add ccode printing for AugmentedAssignment and For (from symcc)

Tests are coming in a later commit.

tokencolour commented:
@asmeurer ,What more needs to be done here?

asmeurer committed:
Add sympy.codegen.ast public names to sympy/codegen/__init__.py

asmeurer committed:
Add tests for ccode for assignments and For

asmeurer committed:
Add fcode printing of For from symcc

asmeurer commented:
This is ready to be reviewed.

bjodah commented on sympy/codegen/ast.py:
```diff
@@ -0,0 +1,391 @@
+     |                                    |--->ModAugmentedAssignment
+     |
+     |--->DataType
+     |           |--------|--->NativeBool
```
Since DataType was dropped for now maybe the docstring shouldn't mention them?

bjodah commented on sympy/codegen/ast.py:
```diff
@@ -0,0 +1,391 @@
+        # causes issues in assignment validation. For now, matrices are defined
+        # as anything with a shape that is not an Indexed
+        lhs_is_mat = hasattr(lhs, 'shape') and not isinstance(lhs, Indexed)
+        rhs_is_mat = hasattr(rhs, 'shape') and not isinstance(rhs, Indexed)
```
`NDimArray` also carries a `shape` property, I'm unsure if that has any implications but I thought I should mention it.
If nothing else renaming the variables `*hs_is_mat` to `*hs_has_shape` is more clear and then treat Indexed as a  special case?

asmeurer replied:
I don't know. This code was just moved from codeprinter.py.

bjodah replied:
I see, NDimArray is newer, that's why then.

bjodah commented on sympy/codegen/ast.py:
```diff
@@ -0,0 +1,391 @@
+    >>> c = CodeBlock(Assignment(x, 1), Assignment(y, x + 1))
+    >>> print(ccode(c))
+    x = 1;
+    y = x + 1;
```
Could we have examples of the mentioned useful methods here?

bjodah commented:
I added some comments, otherwise I am +1 to merge.

In your commit message of bd0a5e7 you say that InOutAssignment is Fortran specific.
Even though it is not directly relevant to this PR, I thought I should share how I view the `intent(InOut)` counterpart in C:

Input arguments in C of say an array may be specified as e.g.:
```c
double do_stuff1(size_t n, const double * const input)
```
if the function is not scalar it may be defined as:
```c
void do_stuff2(size_t n, const double * const restrict input, double * const restrict output)
```
these kinds of things make Fortran syntax quite appealing in comparison, it is very easy to forget to add a `const` qualifier somewhere, hence having something like an `InOut`-intent in SymPy would be useful even in C settings in my humble opinion.

The difference between `InOut` and `Out` would matter if the `CodeBlock` starts with augmented assignments (undefined in the latter case) which would be nice to have `SymPy` catch (even though modern compilers have tools to warn for undefined behavoiur they aren't always perfect).

asmeurer committed:
Remove DataType from the docstring

asmeurer commented:
I see. We will definitely need to have something like this if/when we generate function definitions.

asmeurer committed:
Add CodeBlock to the docstring of sympy.codegen.ast

asmeurer commented:
Going to merge this.
